Register view improvements

Gumball2415 opened the pull request:
- Arrange register layout to 4 bytes per row
- Add modulation pitch and depth
- Arrange waveform view

Gumball2415 committed:
Improve FDS register layout

- Arrange register layout to 4 bytes per row
- Add modulation pitch and depth
- Arrange waveform view

Gumball2415 committed:
Improve FDS register layout

- Arrange register layout to 4 bytes per row
- Add modulation pitch and depth
- Arrange waveform view

Gumball2415 committed:
Fix negative octave bug

Gumball2415 committed:
Fix negative octave bug

Gumball2415 committed:
Justify MIDI note name

Gumball2415 committed:
Justify MIDI note name

Gumball2415 committed:
Improve noise and DPCM register view

- Placed "rate" instead of pitch when sample rate is measured instead
- Justify and align frequency hertz text

Gumball2415 committed:
Improve noise and DPCM register view

- Placed "rate" instead of pitch when sample rate is measured instead
- Justify and align frequency hertz text

Gumball2415 committed:
Justify DPCM register text

Gumball2415 committed:
Justify DPCM register text

Gumball2415 committed:
Add smooth register frequency view setting

Gumball2415 committed:
Add smooth register frequency view setting

Gumball2415 committed:
Implement config setting for smooth frequency view

- Put misc. GUI settings into its own config tab
- Fix warning being treated as error for x86 builds
- Put misc. GUI settings into its own settings category and registry

Gumball2415 committed:
Implement config setting for smooth frequency view

- Put misc. GUI settings into its own config tab
- Fix warning being treated as error for x86 builds
- Put misc. GUI settings into its own settings category and registry

eulyderg committed:
Correct note bar offset + improve FDS wave drawing logic (#120)

eulyderg committed:
Correct note bar offset + improve FDS wave drawing logic (#120)

nyanpasu64 committed:
Fix CMake build

nyanpasu64 committed:
Fix high-res spectrum visualizer at small buffer sizes (#126)

Even with this fix, both the oscilloscope and spectrum visualizer rely
on the visualizer thread processing incoming data faster than the audio
thread can write new data, on a "best-effort" basis. If the visualizer
falls behind, the oscilloscope or spectrum will show a discontinuity in
incoming data. The discontinuity will appear at a random point of the
oscilloscope. In spectrum mode, it will affect the entire spectrum, but
only appears at buffer lengths below 2048 samples.

nyanpasu64 commented on Source/PatternEditor.cpp:
```diff
@@ -1984,35 +2005,53 @@ void CPatternEditor::DrawRegisters(CDC *pDC)
+		CString Modtext;
+
 		FDStext.Format(_T("%s, vol = %02i"), GetPitchTextFunc(3, period, freq), vol);
+		Modtext.Format(_T("modulation %s, depth = %02i"), GetPitchTextFunc(3, modperiod, modfreq), moddepth);
```
I don't like how the "modulation pitch" line is a lot longer than the pitch line, causing the two frequencies to not be aligned vertically for easy reading (and comparing their difference) while the song is playing.

you'd have to change `GetPitchTextFunc` to take a string argument for the name. You could put it at the end of the list, and give it a default value of `_T("pitch")`, to avoid having to change all other uses of the function.

I implemented this locally (https://github.com/nyanpasu64/Dn-FamiTracker/tree/feature-FDS-register-view).

nyanpasu64 commented on Source/PatternEditor.cpp:
```diff
@@ -1832,6 +1836,22 @@ void CPatternEditor::DrawRegisters(CDC *pDC)
+		const int cents = int((note - double(note_conv)) * 100.0);
+
+		CString str;
+		CString ratepitch = rate ? _T("rate ") : _T("pitch");
```
~~instead of `bool rate`, can you pass in a string? (since you're never using `if (rate)` to change functionality, you don't need to know its value, only the string it maps to.) This allows using more names beyond "rate" and "pitch".~~

boolean parameters are not self-documenting and not extensible: https://alexkondov.com/should-you-pass-boolean-to-functions/

Actually keep `bool rate`, it makes sense in context, and you can use this to hide the note name in rate mode if desired.

nyanpasu64 commented on Source/PatternEditor.cpp:
```diff
@@ -1856,14 +1876,15 @@ void CPatternEditor::DrawRegisters(CDC *pDC)
 			period = reg[2] & 0x0F;
 			vol = reg[0] & 0x0F;
-			text.Format(_T("pitch = $%01X, vol = %02i, mode = %i"), period, vol, reg[2] >> 7);
+			text.Format(_T("%s, vol = %02i, mode = %i"), GetPitchTextFuncLong(1, period, freq, !(reg[2] >> 7)), vol, reg[2] >> 7);
```
freq comes from `CSoundGen::GetChannelFrequency` -> `CAPU::GetFreq` -> `C2A03::GetFreq` -> `NES_DMC::GetFrequencyNoise() `. In non-periodic mode it returns a sampling rate (smp/s), in periodic mode it returns a pitch frequency (cycles/s).

It makes no sense to plot a note pitch for `GetPitchTextFuncLong()` to show a music note name corresponding to the sampling rate of a non-periodic noise waveform. Music note names only apply to pitch frequency.

In a local branch (https://github.com/nyanpasu64/Dn-FamiTracker/tree/feature-FDS-register-view), I rewrote the function to hide the music note in rate mode. Is that a good idea, or should I plot the music note anyway?

Gumball2415 replied:
DPCM plots the music note despite it using the sampling rate, should i also rid that too? The noise's sampling rate was based on what DPCM did

Gumball2415 replied:
Merged this feature into the PR

nyanpasu64 committed:
Rename variables

nyanpasu64 committed:
Hide note name in non-pitch noise mode

nyanpasu64 commented on Source/PatternEditor.cpp:
```diff
@@ -1856,14 +1876,15 @@ void CPatternEditor::DrawRegisters(CDC *pDC)
-			text.Format(_T("%s, %s, size = %i byte%c"), GetPitchTextFunc(1, period & 0x0F, freq),
-				(reg[0] & 0x40) ? _T("looped") : _T("once"), (reg[3] << 4) | 1, reg[3] ? 's' : ' ');
+			vol = 15 *!pSoundGen->PreviewDone();
+			text.Format(_T("%s, %-5s size = %-4i byte%c"), GetPitchTextFuncLong(1, (period & 0x0F), freq, 1),
```
It would be cool if, in DPCM mode, sufficiently short looped samples had their periodicities plotted. Not sure if this will suffer from wrong frequencies so often it's useless in practice. For example, https://www.youtube.com/watch?v=v-gdIjKd7Ic has a looped sample of a high note, but the loop length is like 17 wave periods long or something, so the register viewer shows a period 4 octaves down at the wrong note, and every DPCM sample has a *differently* wrong loop length.

Will a DPCM pitch viewer be right or wrong more often, or not show up at all in nearly all modules?

nyanpasu64 committed:
Align FDS modulator text with carrier

nyanpasu64 commented on Dn-FamiTracker.rc:
```diff
@@ -2020,6 +2029,25 @@ BEGIN
+    LTEXT           "Channel view limit",IDC_STATIC,14,17,58,8
+    LTEXT           "Idle refresh rate",IDC_STATIC,14,52,54,8
+    CONTROL         "",IDC_IDLE_REFRESH_SLIDER,"msctls_trackbar32",TBS_BOTH | TBS_NOTICKS | WS_TABSTOP,133,50,91,12
+    CONTROL         "Smooth register frequency view",IDC_SMOOTHREGFREQ_CHECK,
```
I found the new GUI tab in the settings, and thought this would change the pitch register or frequency *text* displays. It actually affects the quantization of the per-channel pitch bar graph.

Ideas:

- Fine pitch {bar/bars/gauge/meter/plot/indicator}
- Fine-grained? Smooth? Precise? Sub-note?

idk.

"Precise pitch bar" sounds half decent.

Gumball2415 replied:
i've given it some thought and wondered if there should be a toggle switch for the text verbosity (i.e. only displaying the registers and channel data, no datatype displayed such as "pitch =" or "size =")

nyanpasu64 committed:
[wip] Rename GUI setting to "Precise pitch bar"

Does not change name in registry!

Gumball2415 committed:
Merge branch 'feature-FDS-register-view' of https://github.com/Dn-Programming-Core-Management/Dn-FamiTracker into feature-FDS-register-view

Gumball2415 committed:
Rename registry setting to PreciseRegPitch

Gumball2415 committed:
Refactor FDS register view

 - Add wavetable register view
 - Add modulation register data
 - Add output modulation pitch in volume row

Gumball2415 committed:
Refactor FDS register view more

 - Use a filled wide rectangle instead of an extra colored rectangle
 - Get output frequency directly from FDS emulation
